src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/pullrequest/PullRequestAwareSCMHeadObserver.java

```diff
@@ -0,0 +1,63 @@
+package com.atlassian.bitbucket.jenkins.internal.scm.pullrequest;
+
+import jenkins.scm.api.SCMHead;
+import jenkins.scm.api.SCMHeadObserver;
+import jenkins.scm.api.SCMRevision;
+import jenkins.scm.api.mixin.ChangeRequestSCMHead;
+
+import java.io.IOException;
+import java.util.Collection;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Custom {@link SCMHeadObserver} that converts {@link SCMRevision revisions} into
+ * {@link BitbucketPullRequestSCMRevision pull request revisions} so that the {@link SCMHead} is able to recognize
+ * them as such and create the "merge commit" to be used for the build
+ */
+public class PullRequestAwareSCMHeadObserver extends SCMHeadObserver {
+
+    private final SCMHeadObserver delegate;
+    private final PullRequestRevisionRetriever revisionRetriever;
+
+    public PullRequestAwareSCMHeadObserver(SCMHeadObserver delegate,
+                                           PullRequestRevisionRetriever revisionRetriever) {
+        this.delegate = requireNonNull(delegate, "delegate");
+        this.revisionRetriever = requireNonNull(revisionRetriever, "revisionRetriever");
+    }
+
+    @Override
+    public void observe(SCMHead head, SCMRevision revision) throws IOException, InterruptedException {
+        if (head instanceof ChangeRequestSCMHead) {
+            // If the head is already a ChangeRequestSCMHead then we simply pass it to the delegate observer
+            unwrapAndObserve(head, revision);
+            return;
+        }
+
+        // Otherwise, we check if there are open pull requests for the specified head and convert the head and revision
+        // into BitbucketChangeRequestSCMHead and BitbucketChangeRequestSCMRevision respectively and pass those to the
+        // delegate observer so they can be build appropriately
+        Collection<BitbucketPullRequestSCMRevision> prRevisions = revisionRetriever.getPullRequestRevisions(head);
+        if (!prRevisions.isEmpty()) {
+            for (BitbucketPullRequestSCMRevision prRevision : prRevisions) {
+                unwrapAndObserve(prRevision.getHead(), prRevision);
```

Comment:
Ideally this should be done in the `retrieve` method, but as a workaround, we are simply transforming the head and revision into their change request equivalents before they are sent to be built.